test/cpp/end2end/client_lb_end2end_test.cc

```diff
@@ -2251,6 +2251,9 @@ class ClientLbInterceptTrailingMetadataTest : public ClientLbEnd2endTest {
     return std::move(load_report_);
   }
 
+  grpc::internal::Mutex mu_;
```

Comment:
Please make these data members private, and add a method to wait for the next event:

```
void WaitForLbCallback() {
  grpc::internal::MutexLock lock(&mu_);
  cond_.WaitWithTimeout(&mu_, absl::Seconds(3));
}
```

Reply:
Thinking about this further, a more general solution might be to have a queue of LB callback data and then have a method to get the latest result within some specified timeout.  For example:

```
public:
  absl::optional<LbTrailingMetadataCallbackData> GetNextTrailingMetadataCallbackData(absl::Duration timeout) {
    grpc::internal::MutexLock lock(&mu_);
    if (queue_.empty()) {
      grpc::internal::CondVar cv;
      cv_ = cv;
      if (!cv.WaitWithTimeout(&mu_, timeout)) return absl::nullopt;
    }
    auto result = std::move(queue_.front());
    queue_.pop_front();
    return result;
  }

 private:
  struct LbTrailingMetadataCallbackData { ... };

  static void ReportTrailerIntercepted(
      const grpc_core::TrailingMetadataArgsSeen& args_seen) {
    LbTrailingMetadataCallbackData entry = // ...convert from args_seen ...
    ClientLbInterceptTrailingMetadataTest* self = current_test_instance_;
    grpc::internal::MutexLock lock(&self->mu_);
    queue_.push_back(std::move(entry));
    if (cv_ != nullptr) {
      cv_->Signal();
      cv_ = nullptr;
    }
  }

  grpc::internal::Mutex mu_;
  grpc::internal::CondVar* cv_ ABSL_GUARDED_BY(&mu_) = nullptr;
  std::deque<LbTrailingMetadataCallbackData> queue_ ABSL_GUARDED_BY(&mu_);
```

This way, we're not reusing the same `CondVar` for multiple LB callbacks.

Reply:
Hm so I have run the test many times and I'm pretty sure this test case is the only place where this race is happening - I'm wondering if we need to add this complexity? I updated with the previous review comment and the test is passing in 100k runs

Reply:
I was thinking that using the same `CondVar` for multiple LB callbacks might cause problems.  If it's not causing problems for any of the other tests, then this seems fine as-is.

Reply:
Yeah, this is the only test in the suite that has this issue, I haven't seen it cause any issues - will run with blase once more to make sure.

Reply:
Ok ran with blase again - no failures for this test suite.